features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/LinkdTopologyBundleActivator.java

```diff
@@ -0,0 +1,121 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2012-2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.plugins.topo.linkd.internal;
+
+import java.util.ArrayList;
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+
+import org.opennms.features.topology.api.support.VertexHopGraphProvider;
+import org.opennms.features.topology.api.topo.DefaultTopologyProviderInfo;
+import org.opennms.features.topology.api.topo.MetaTopologyProvider;
+import org.opennms.features.topology.api.topo.SimpleMetaTopologyProvider;
+import org.opennms.features.topology.api.topo.TopologyProviderInfo;
+import org.opennms.netmgt.topologies.service.api.OnmsTopologyDao;
+import org.opennms.netmgt.topologies.service.api.OnmsTopologyProtocol;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+
+public class LinkdTopologyBundleActivator {
```

Comment:
I am not sure what you try to achieve here.
We should expose this via blueprint instead.

Reply:
I do not know how to do.
 I need to cycle over protocols.

Reply:
I mean I do not know how to cycle inside blueprint.

Reply:
Yeah, my assumption was, that we are using the MetaTopologyProvider, which I still think we should be using instead of exposing all the protocols.

Reply:
I see. But I prefer to have all the supported protocol listed in the Menu. They are not so much so we can reasonably have this.
Also I think that the "LinkdTopologyProvider" could provide wrong topology in the case CDP is enabled.

Reply:
Markus I got the maps.....and the best would be to have  a unique MetaTopologyProvider with the ability to select the protocols.....so you pass the protocol list to the LinkdTopolologyProvider and everything i done....
But I have no idea of how we can do this....
I remember you wrote some code....but the best would be register a menu entry that is selectable for every register protocol.....So I have one LinkdTopologyProvider and I can check the protocol to be displayed.

Reply:
Do you have an idea of how I can do this?
Example code?

Reply:
I get the protocol maps..and they are not right...because for example OSPF-MIB is not enabled on the router. And enlinkd is not able to get ospf info but maybe has CDP info.....
So I need to have enabled all the protocols as checked box.

Further more I guess I'm going to add an iplike search provider....would be useful if you want to display all the nodes inside a network

Reply:
I have done the required changes....just committed the latest changes...

commit 2d355a4b830dfe454198206188ce9b9762dda496 (HEAD -> jira/NMS-12410, origin/jira/NMS-12410)
Author: Antonio <rssntn67@yahoo.it>
Date:   Fri Dec 6 14:11:20 2019 +0100

    Removed Bundle Activator

    Default Topology Provider is on constructor

commit e0d75f796ffb0ef32e76bbb0d0beabb25556f8ec
Author: Antonio <rssntn67@yahoo.it>
Date:   Fri Dec 6 12:35:59 2019 +0100

    NMS-12410: Introduce getNamespace for SearchQuery

    This has been done by mvr in a previous commit.

commit 4e83316a3af7709c95fe1ccc15db34655a09ff2e
Author: Antonio <rssntn67@yahoo.it>
Date:   Fri Dec 6 11:55:41 2019 +0100

    enabling both Activator and MetaTopologyProvider

    Search does not work on LinkdMetaTopologyProvider

commit 6f9f1d7c8f8540fe779bfede4a965367f22bb01b
Author: Antonio <rssntn67@yahoo.it>
Date:   Thu Dec 5 18:57:13 2019 +0100

    Readded only LinkdMetaTopologyProvider

Good work Markus